packages/protocol/contracts/L1/libs/LibProposing.sol

```diff
@@ -151,145 +113,209 @@ library LibProposing {
-        if (blk.blockId != blockId) {
-            revert L1_INVALID_BLOCK_ID();
-        }
+        return keccak256(
+            abi.encode(
+                "PROVER_ASSIGNMENT",
+                txListHash,
+                assignment.feeToken,
+                assignment.expiry,
+                assignment.tierFees
+            )
+        );
     }
 
-    function _validateBlock(
-        TaikoData.State storage state,
-        TaikoData.Config memory config,
-        TaikoData.BlockMetadataInput memory input,
-        bytes calldata txList
+    function _validateAssignment(
+        uint16 minTier,
+        bytes32 txListHash,
+        TaikoData.ProverAssignment memory assignment
     )
         private
-        view
-        returns (bool cacheTxListInfo)
+        returns (uint256 proverFee)
     {
-        if (input.proposer == address(0)) revert L1_INVALID_METADATA();
+        // Check assignment not expired
+        if (block.timestamp >= assignment.expiry) {
+            revert L1_ASSIGNMENT_EXPIRED();
+        }
 
-        // handling txList
+        if (txListHash == 0 || assignment.prover == address(0)) {
+            revert L1_ASSIGNMENT_INVALID_PARAMS();
+        }
 
-        uint24 size = uint24(txList.length);
-        if (size > config.blockMaxTxListBytes) revert L1_TX_LIST();
+        // Hash the assignment with the txListHash, this hash will be signed by
+        // the prover, therefore, we add a string as a prefix.
+        bytes32 hash = hashAssignmentForTxList(assignment, txListHash);
 
-        if (input.txListByteStart > input.txListByteEnd) {
-            revert L1_TX_LIST_RANGE();
+        if (!assignment.prover.isValidSignature(hash, assignment.signature)) {
+            revert L1_ASSIGNMENT_INVALID_SIG();
         }
 
-        if (config.blockTxListExpiry == 0) {
-            // caching is disabled
-            if (input.txListByteStart != 0 || input.txListByteEnd != size) {
-                revert L1_TX_LIST_RANGE();
-            }
-        } else {
-            // caching is enabled
-            if (size == 0) {
-                // This blob shall have been submitted earlier
-                TaikoData.TxListInfo memory info =
-                    state.txListInfo[input.txListHash];
-
-                if (input.txListByteEnd > info.size) {
-                    revert L1_TX_LIST_RANGE();
-                }
-
-                if (
-                    info.size == 0
-                        || info.validSince + config.blockTxListExpiry
-                            < block.timestamp
-                ) {
-                    revert L1_TX_LIST_NOT_EXIST();
-                }
-            } else {
-                if (input.txListByteEnd > size) revert L1_TX_LIST_RANGE();
-                if (input.txListHash != keccak256(txList)) {
-                    revert L1_TX_LIST_HASH();
-                }
+        // Find the prover fee using the minimal tier
+        proverFee = _getProverFee(assignment.tierFees, minTier);
```

Comment:
Shall we require `assignment.tierFees.length == ITierProvider(resolve("tier_provider", false)).getTierIds().length` here?

Reply:
I don't think the check is necessary either.